// vercel-php/dist/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.readRuntimeFile = exports.ensureLocalPhp = exports.getComposerFiles = exports.getLauncherFiles = exports.modifyPhpIni = exports.getPhpFiles = exports.getIncludedFiles = void 0;
const path_1 = __importDefault(require("path"));
const child_process_1 = require("child_process");
const build_utils_1 = require("@vercel/build-utils");
const libphp = __importStar(require("@libphp/amazon-linux-2-v74"));
const PHP_PKG = path_1.default.dirname(require.resolve('@libphp/amazon-linux-2-v74/package.json'));
const PHP_BIN_DIR = path_1.default.join(PHP_PKG, "native/php");
const PHP_MODULES_DIR = path_1.default.join(PHP_BIN_DIR, "modules");
const PHP_LIB_DIR = path_1.default.join(PHP_PKG, "native/lib");
const COMPOSER_BIN = path_1.default.join(PHP_BIN_DIR, "composer");
async function getIncludedFiles({ files, entrypoint, workPath, config, meta }) {
    // Download all files to workPath
    const downloadedFiles = await build_utils_1.download(files, workPath, meta);
    let includedFiles = {};
    if (config && config.includeFiles) {
        // Find files for each glob
        for (const pattern of config.includeFiles) {
            const matchedFiles = await build_utils_1.glob(pattern, workPath);
            Object.assign(includedFiles, matchedFiles);
        }
        // explicit and always include the entrypoint
        Object.assign(includedFiles, {
            [entrypoint]: files[entrypoint],
        });
    }
    else {
        // Backwards compatibility
        includedFiles = downloadedFiles;
    }
    return includedFiles;
}
exports.getIncludedFiles = getIncludedFiles;
async function getPhpFiles({ meta }) {
    const files = await libphp.getFiles();
    if (meta && meta.isDev) {
        delete files['php/php'];
        delete files['php/php-fpm'];
        delete files['php/php-fpm.ini'];
    }
    else {
        delete files['php/php-cgi'];
        delete files['php/php-fpm'];
        delete files['php/php-fpm.ini'];
    }
    const runtimeFiles = {};
    for (const [filename, filepath] of Object.entries(files)) {
        runtimeFiles[filename] = new build_utils_1.FileFsRef({
            fsPath: filepath
        });
    }
    return runtimeFiles;
}
exports.getPhpFiles = getPhpFiles;
function modifyPhpIni(phpini, directives) {
    const output = [];
    for (const property in directives) {
        output.push(`${property} = ${directives[property]}`);
    }
    phpini.data = phpini.data
        .toString()
        .concat(output.join("\n"));
    return phpini;
}
exports.modifyPhpIni = modifyPhpIni;
function getLauncherFiles({ meta }) {
    const files = {
        'helpers.js': new build_utils_1.FileFsRef({
            fsPath: path_1.default.join(__dirname, 'launchers/helpers.js'),
        })
    };
    if (meta && meta.isDev) {
        files['launcher.js'] = new build_utils_1.FileFsRef({
            fsPath: path_1.default.join(__dirname, 'launchers/cgi.js'),
        });
    }
    else {
        files['launcher.js'] = new build_utils_1.FileFsRef({
            fsPath: path_1.default.join(__dirname, 'launchers/builtin.js'),
        });
    }
    return files;
}
exports.getLauncherFiles = getLauncherFiles;
async function getComposerFiles(workPath) {
    console.log('ðŸ˜ Installing Composer deps.');
    // Install composer dependencies
    await runComposerInstall(workPath);
    console.log('ðŸ˜ Installing Composer deps OK.');
    return await build_utils_1.glob('vendor/**', workPath);
}
exports.getComposerFiles = getComposerFiles;
async function runComposerInstall(cwd) {
    // @todo think about allow to pass custom commands here
    await runPhp(cwd, [
        COMPOSER_BIN,
        'install',
        '--profile',
        '--no-dev',
        '--no-interaction',
        '--no-scripts',
        '--ignore-platform-reqs',
        '--no-progress'
    ], { stdio: 'inherit' });
}
async function runPhp(cwd, args, opts = {}) {
    try {
        await spawnAsync('php', [`-dextension_dir=${PHP_MODULES_DIR}`, ...args], cwd, {
            ...opts,
            ...{
                env: {
                    ...process.env,
                    ...{
                        COMPOSER_HOME: '/tmp',
                        PATH: `${PHP_BIN_DIR}:${process.env.PATH}`,
                        LD_LIBRARY_PATH: `${PHP_LIB_DIR}:/usr/lib64:/lib64:${process.env.LD_LIBRARY_PATH}`
                    }
                }
            }
        });
    }
    catch (e) {
        console.error(e);
        process.exit(1);
    }
}
async function ensureLocalPhp() {
    try {
        await spawnAsync('which', ['php', 'php-cgi'], undefined, { stdio: 'pipe' });
        return true;
    }
    catch (e) {
        return false;
    }
}
exports.ensureLocalPhp = ensureLocalPhp;
async function readRuntimeFile(file) {
    const blob = await build_utils_1.FileBlob.fromStream({
        stream: file.toStream(),
    });
    return blob.data.toString();
}
exports.readRuntimeFile = readRuntimeFile;
function spawnAsync(command, args, cwd, opts = {}) {
    return new Promise((resolve, reject) => {
        const child = child_process_1.spawn(command, args, {
            stdio: "ignore",
            cwd,
            ...opts
        });
        child.on('error', reject);
        child.on('exit', (code, signal) => {
            if (code === 0) {
                resolve();
            }
            else {
                reject(new Error(`Exited with ${code || signal}`));
            }
        });
    });
}
